### PR TITLE
storage capacity: object ownership

### LINKED ISSUE #465: storage capacity: make ownership of CSIStorageCapacity objects optional
Apparently some CSI drivers run outside of the cluster.

https://github.com/kubernetes-csi/external-provisioner/pull/450/files#r470300935

### PR DESCRIPTION
**Pull request by pohly**
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:

This promotes "storage capacity tracking" to beta and changes its usages as agreed upon during KEP review for beta (optional ownership, labels used to identify objects managed by each instance).

**Which issue(s) this PR fixes**:

Fixes #465 

**Special notes for your reviewer**:

Based on https://github.com/kubernetes-csi/external-provisioner/pull/579

**Does this PR introduce a user-facing change?**:
```release-note
Storage capacity tracking becomes a beta feature. The parameters are different (`POD_NAMESPACE` -> `NAMESPACE`) and `CSIStorageCapacity` objects created via the alpha feature in previous releases must be deleted before rolling out the new release, because the way how each instance identifies the objects to manage has changed (from owner reference to labels).
```

### CODE REVIEW

**Comment by pohly** [comment]
/hold

Let's merge this one last once all other changes for storage capacity tracking are in.

**Comment by verult** [comment]
Just curious: how does a label selector increase API server CPU load?

File: cmd/csi-provisioner/csi-provisioner.go
Comment (verult) [inline cmd/csi-provisioner/csi-provisioner.go:469]: I'm wondering how easy it is to detect if the informer doesn't fetch objects we expect, and whether we need to unit test this.

If informer logic is wrong, two things could happen:

* Missing items that are supposed to be included
  * May try to create an existing CSIStorageCapacity object. In the logic today this is fine, because nothing is done after object creation (besides logging / error reporting), and the creation error would give us a hint.
* Includes more things than it should
  * No failures, just more network traffic.

The latter case is hard to detect. So maybe a unit test is worthwhile (if it's not too much effort)?

**Comment by pohly** [comment]
> Just curious: how does a label selector increase API server CPU load?

I had exactly the same concern and asked on #sig-scalability. The response was that spending more CPU time on server-side filtering can be assumed to be a net win when it decreases the number of objects that need to be transferred. That costs both network bandwidth *and* CPU time for conversion.

**Comment by chrishenzie** [comment]
When you have a moment can you please address this merge conflict? We are planning to release this tomorrow to align with the Kubernetes 1.21 release.

File: cmd/csi-provisioner/csi-provisioner.go
Comment (verult) [reply cmd/csi-provisioner/csi-provisioner.go:469]: Actually since this is just an options configuration, the logic is simple enough, so it's probably low risk. Totally up to you if you want to unit test this.

**Comment by verult** [comment]
> I had exactly the same concern and asked on #sig-scalability. The response was that spending more CPU time on server-side filtering can be assumed to be a net win when it decreases the number of objects that need to be transferred. That costs both network bandwidth and CPU time for conversion.

Since I'm not familiar with API server code, I was actually wondering exactly how computationally heavy server-side filtering is. What does it do under the hood to do this filtering?

File: cmd/csi-provisioner/csi-provisioner.go
Comment (verult) [inline cmd/csi-provisioner/csi-provisioner.go:96]: 1 for a StatefulSet *or DaemonSet*

File: pkg/capacity/capacity.go
Comment (verult) [inline pkg/capacity/capacity.go:565]: Should this be `isManaged()`? If a driver running in binary mode manages an object which is unchanged, it should still skip right?

File: pkg/capacity/capacity.go
Comment (verult) [inline pkg/capacity/capacity.go:748]: Since the informer is already filtered by labels in the next commit, and all callers of `isManaged()` fetch from the informer, is this mostly here as a safeguard? If so, nit: could you add a comment to clarify?

File: pkg/capacity/capacity.go
Comment (verult) [inline pkg/capacity/capacity.go:601]: nit: update function name in comment

File: pkg/capacity/capacity_test.go
Comment (verult) [inline pkg/capacity/capacity_test.go:1267]: Should there be a test case for "managed by us but not owned by us"?

**Review by verult — CHANGES_REQUESTED** [review]

File: cmd/csi-provisioner/csi-provisioner.go
Comment (pohly) [reply cmd/csi-provisioner/csi-provisioner.go:453]: I'd prefer to not test this. I suspect that it wouldn't be easy to test comprehensively and in the end is more about correct implementation of the informer and apiserver filtering than it is about correct configuration.

File: README.md
Comment (pohly) [reply README.md:163]: Added.

As you said, orphaned and incorrect objects cause no permanent damage for pods with normal PVCs (which includes generic ephemeral volumes), so it's not too bad. It's worse for CSI ephemeral volumes because then a pod might get scheduled onto a node where the driver then cannot provide the volume and the pod is stuck there (no rescheduling). I think that was what we chatted about, not the scheduler itself hanging.

File: README.md
Comment (pohly) [reply README.md:168]: Added.

File: cmd/csi-provisioner/csi-provisioner.go
Comment (pohly) [reply cmd/csi-provisioner/csi-provisioner.go:96]: added

File: pkg/capacity/capacity.go
Comment (pohly) [reply pkg/capacity/capacity.go:640]: This really is an ownerref check. It covers the scenario where a driver was deployed without ownerref and then gets redeployed with ownerref. The existing objects then need to be updated even when the capacity is the same, which is achieved here by falling through to the final else clause.

I added "same capacity %v *and correct owner*" to the log message to make this more clear.

File: pkg/capacity/capacity.go
Comment (pohly) [reply pkg/capacity/capacity.go:823]: Correct. I added a comment.

File: pkg/capacity/capacity.go
Comment (pohly) [reply pkg/capacity/capacity.go:677]: Fixed.

File: pkg/capacity/capacity_test.go
Comment (pohly) [reply pkg/capacity/capacity_test.go:1273]: There is "reuse one capacity object, add owner" for that.

Comment (pohly) [reply pkg/capacity/capacity_test.go:1273]: The inverse "reuse one capacity object, remove owner" doesn't work because the driver instance without owner configured doesn't know which owner references it is expected to remove. This shouldn't be needed in practice and the documentation mentions it.

**Comment by pohly** [comment]
> Since I'm not familiar with API server code, I was actually wondering exactly how computationally heavy server-side filtering is. What does it do under the hood to do this filtering?

It has an in-memory representation of the object and then simply checks labels in the meta data. In that sense it's simply brute force. But the in-memory representation is needed anyway and then the check is still faster than the rest of the processing that would need to happen when sending the object. At least that is my understanding...

**Comment by pohly** [comment]
I pushed an updated version. As before it is based on PR #586

**Comment by pohly** [comment]
> When you have a moment can you please address this merge conflict? We are planning to release this tomorrow to align with the Kubernetes 1.21 release.

What about updating client-go in external-provisioner to 1.21.0? Kubernetes 1.21 must be released first, then I can update client-go and only then can we release external-provisioner.

Or we release external-provisioner with the current code. None of the differences seem relevant. I created the update for rc0 here: https://github.com/kubernetes-csi/external-provisioner/pull/599

**Comment by chrishenzie** [comment]
@pohly I had planned on updating to 1.21 when it was released. Thank you for taking care of this.

**Comment by verult** [comment]
> It has an in-memory representation of the object and then simply checks labels in the meta data. In that sense it's simply brute force. But the in-memory representation is needed anyway and then the check is still faster than the rest of the processing that would need to happen when sending the object. At least that is my understanding...

Yeah that makes sense. Some sort of index could maybe speed up processing but maybe that's overkill 🤷

File: README.md
Comment (verult) [reply README.md:163]: Ah yeah. Could you maybe mention this in the design doc somewhere (if it's not there already)?

File: pkg/capacity/capacity.go
Comment (verult) [inline pkg/capacity/capacity.go:640]: I see, and `c.owner == nil` covers the case that a driver stays owner-less. Sgtm!

File: pkg/capacity/capacity_test.go
Comment (verult) [reply pkg/capacity/capacity_test.go:1273]: I was actually referring to a test that captures the container-less driver case, where ownerref stays nil always. IIUC in all the tests so far assumes that capacity objects are eventually managed by a driver in containers (i.e. owner is always set to `defaultOwner`).

If you agree this test is necessary I would prefer that the test is added in this PR rather than delaying it after the release, because it is a use case we plan to support. Hopefully it's not a lot of work.

**Review by verult — CHANGES_REQUESTED** [review]
One more potential change, everything else lgtm

File: README.md
Comment (pohly) [reply README.md:163]: It is covered by the generic ephemeral volume KEP.

File: pkg/capacity/capacity_test.go
Comment (pohly) [reply pkg/capacity/capacity_test.go]: You are right, that is a supported use case and wasn't covered. I've added two test cases for it which cover the relevant difference (create objects without owner reference, keep existing references).

**Comment by verult** [comment]
/lgtm

/assign @msau42 
For approval

File: README.md
Comment (msau42) [inline README.md:79]: Update this to also include Daemonset

**Review by msau42 — COMMENTED** [review]
/approve

just some nits

**Comment by pohly** [comment]
> just some nits

Both fixed, please lgtm again.

/hold cancel

**Pull request merged**
